http.go

```diff
@@ -106,6 +116,45 @@ func InstrumentHandlerFunc(label string, f func(http.ResponseWriter, *http.Reque
 	return InstrumentHandler(label, http.HandlerFunc(f))
 }
 
+// traceSampler is a time-based request sampler for tracing.
+//
+// It allows at most one request to be traced during each time interval.
+type traceSampler struct {
+	interval time.Duration
+	rng      *rand.Rand
+
+	mu             sync.Mutex
+	nextSampleTime time.Time
+}
+
+// newTraceSampler returns a new traceSampler that allows at most one request
+// to be traced during each time interval.
+func newTraceSampler(interval time.Duration) *traceSampler {
+	now := time.Now()
+	return &traceSampler{
+		interval:       interval,
+		rng:            rand.New(rand.NewSource(now.UnixNano())),
+		nextSampleTime: now,
+	}
+}
+
+// shouldTrace returns true iff a request should be traced.
+func (s *traceSampler) shouldTrace() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	now := time.Now()
+	if now.Before(s.nextSampleTime) {
+		return false
+	}
+
+	// Pick a new value for nextSampleTime and return.
+	const jitter = 0.4
+	rv := 2*s.rng.Float64() - 1.0 // range [-1.0, 1.0)
+	mult := 1 + jitter*rv
+	s.nextSampleTime = s.nextSampleTime.Add(time.Duration(float64(s.interval) * mult))
```

Comment:
I think this needs to be:

```
s.nextSampleTime = now.Add(...)
```

That way if we ever have a long idle period, we won't sample a lot of requests in the "past".

Reply:
Good point, done!